### PR TITLE
Major rework of shutdown and reconnect logs to avoid zombie tasks and race conditions

### PR DESCRIPTION
**Pull request by Drapersniper**
Depends on #115 
~~Depends on #118  (For the callback code)~~ Moved the callback code to the logging PR.
Depends on #119

Merge Last

### CODE REVIEW

File: lavalink/lavalink.py [hunk 1]
```diff
@@ -378,3 +378,29 @@ def all_connected_players() -> Tuple[player_manager.Player]:
...
+    ]
+    if tasks:
+        done, pending = await asyncio.wait(
+            tasks, timeout=timeout, return_when=asyncio.ALL_COMPLETED
+        )
+        for task in pending:
+            task.cancel()
+        if done:
+            for task_result in done:
+                task_result.result()
```
Comment (Jackenmen) [inline lavalink/lavalink.py]: `asyncio.gather()` could be used here instead.
```suggestion
    for result in await asyncio.gather(
        *(
            node_.wait_until_ready(timeout)
            for node_ in node._nodes
            if (not node_.ready) and not node_.state == node.NodeState.DISCONNECTING
        ),
        return_exceptions=True,
    ):
        if result is not None:
            raise result
```
This *could* be wrapped in `asyncio.wait_for()` but I don't think that's really needed considering you already pass a timeout argument to `wait_until_ready()`.

File: lavalink/node.py [hunk 2]
```diff
@@ -226,34 +233,30 @@ async def connect(self, timeout=None):
...
-        self._ready_event.set()
-        self.update_state(NodeState.READY)
+        if self.try_connect_task is not None:
+            self.try_connect_task.cancel()
+        try:
+            self.try_connect_task = asyncio.create_task(self._multi_try_connect(uri))
+            for task in asyncio.as_completed([self.try_connect_task], timeout=timeout):
+                await task
+        except asyncio.CancelledError:
+            raise AbortingNodeConnection
```
Comment (Jackenmen) [inline lavalink/node.py]: Let's not overengineer these before it's actually needed for something.
```suggestion
        self.try_connect_task = asyncio.create_task(self._multi_try_connect(uri))
        try:
            await asyncio.wait_for(self.try_connect_task, timeout=timeout)
        except asyncio.CancelledError:
            raise AbortingNodeConnection
```

File: lavalink/lavalink.py [hunk 3]
```diff
@@ -378,3 +378,29 @@ def all_connected_players() -> Tuple[player_manager.Player]:
 def active_players() -> Tuple[player_manager.Player]:
     ps = all_connected_players()
     return tuple(p for p in ps if p.is_playing)
+
+
+async def wait_until_ready(timeout: Optional[float] = None, wait_if_no_node: Optional[int] = None):
+
```
Comment (Jackenmen) [inline lavalink/lavalink.py]:
```suggestion
```

File: lavalink/node.py [hunk 4]
```diff
@@ -226,34 +233,29 @@ async def connect(self, timeout=None):
...
             ``None`` is considered never, but the underlying code may stop trying past a
             certain point.
-
+        shutdown: bool
+            Whether the node was told to shut down
         Raises
         ------
         asyncio.TimeoutError
             If the websocket failed to connect after the given time.
+        AbortingConnectionException:
```
Comment (Drapersniper) [inline lavalink/node.py]:
```suggestion
        AbortingConnection:
```

**Review by aikaterna — APPROVED** [review]

**Pull request merged**